src/Membership.sol

```diff
@@ -0,0 +1,427 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+import { IPriceCalculator } from "./IPriceCalculator.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import "openzeppelin-contracts/contracts/utils/Context.sol";
+
+// The number of periods should be greater than zero
+error NumberOfPeriodsCantBeZero();
+
+// ETH Amount passed as value on the transaction is not correct
+error IncorrectAmount();
+
+// An eth value was assigned in the transaction and only tokens were expected
+error OnlyTokensAccepted();
+
+// The specified rate limit was not correct or within the expected limits
+error InvalidRateLimit();
+
+// It's not possible to acquire the rate limit due to exceeding the expected limits
+// even after attempting to erase expired memberships
+error ExceedMaxRateLimitPerEpoch();
+
+// This membership is not in grace period yet
+error NotInGracePeriod(uint256 idCommitment);
+
+// The sender is not the holder of the membership
+error NotHolder(uint256 idCommitment);
+
+// This membership cannot be erased (either it is not expired or not in grace period and/or not the owner)
+error CantEraseMembership(uint256 idCommitment);
+
+contract Membership {
+    using SafeERC20 for IERC20;
+
+    /// @notice Address of the Price Calculator used to calculate the price of a new membership
+    IPriceCalculator public priceCalculator;
+
+    /// @notice Maximum total rate limit of all memberships in the tree
+    uint32 public maxTotalRateLimitPerEpoch;
```

Comment:
This would let us have a maximum total rate limit per epoch of `4_294_967_295`. This should be large enough. if we change it to `uint16` we can potentially use a maximum ratelimit of `65_535` messages per epoch. Is this value large enough?

Reply:
> use a maximum ratelimit of 65_535 messages per epoch. Is this value large enough?

Considering the memory usage to keep track. I believe `uint16` is large enough. See https://github.com/waku-org/specs/blob/master/standards/core/rln-contract.md#why-this-particular-epoch-length

Reply:
Hm. I forgot that in the spec the max is set to 160_000 which exceeds the max of uint16